### PR TITLE
Enable status-aware redirect balancing in the Director

### PR DESCRIPTION
**Pull request by jhiemstrawisc**
This new section of code uses servers' reported status as part of the
sorting algorithm when the Director enables "adaptive" sort.

It works by starting with a raw weight 'x' in (0,1] that corresponds
to different statuses. From here, x at time t (xt) is adjusted to
take into account past known weights using an EWMA low-pass filter,
resulting in st. This smoothing function is applied to avoid sharp
box-wave-like changes in the server's priority due to status changes.

Effectively, this penalizes a server's priority proportional to the
length of time the server spends outside of the "ok" state. Perhaps
the most useful non-ok status to think about is the "degraded" status,
which occurs for servers that exceed some threshold relative to their
configured IO throttle limit. This means that servers under heavy load
(as defined by percentage of configured allowable connections) will
be slowly deprioritized for client redirects until the server has a
chance to dip below the configured limit.

### CODE REVIEW

File: director/cache_ads.go [hunk 1]
```diff
@@ -105,6 +107,108 @@ func SetBrokerDialer(dialer *broker.BrokerDialer) {
...
+
+	st_1 := oldSAd.StatusWeight
+	// If we received an out-of-bounds previous status weight,
+	// reset it to 1 -- This points to something having gone wrong,
+	// so the safest choice is to avoid affecting the server's sorting.
+	if st_1 <= 0 || st_1 > 1 {
+		st_1 = 1.0
+	}
+	t_1 := oldSAd.StatusWeightLastUpdate
+	deltaT := t.Sub(time.Unix(t_1, 0))
```
[the remark this answers is not in the document]
Comment (jhiemstrawisc) [reply director/cache_ads.go:181]: And I'm adding a unit test for negative a negative delta T to make sure the derived weight is correctly imputed.

**Review by patrickbrophy — APPROVED** [review]
Thanks for implementing the changes! I can see that the metrics are updating as the status changes in the service. Hopefully we can get this turned on soon.
